### ISSUE TITLE
[SYCL] Improve traceability of loading libraries

### ISSUE THREAD

**Issue by marekkozl** [intel/llvm#4371]
This is a very simple request to add dlerror message in case of dlopen fails. It can be enabled via specific level of tracing. 
Adding custom layer of Level Zero loader might not be initialized properly and the device won't be detectable without any feedback from SYCL runtime.

**Comment by vladimirlaz** [comment]
@smaslov-intel could you comment on the issue?

**Comment by alexbatashev** [comment]
Isn't `SYCL_PI_TRACE=-1` already doing this?

**Comment by vladimirlaz** [comment]
AFAIK it does not provide information about loaded libraries.

**Comment by alexbatashev** [comment]
@vladimirlaz 
```bash
$ SYCL_PI_TRACE=-1 ./build/bin/sycl-ls
SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_opencl.so
SYCL_PI_TRACE[basic]: Plugin found and successfully loaded: libpi_level_zero.so
SYCL_PI_TRACE[all]: Check if plugin is present. Failed to load plugin: libpi_cuda.so
SYCL_PI_TRACE[all]: Check if plugin is present. Failed to load plugin: libpi_hip.so
SYCL_PI_TRACE[all]: Check if plugin is present. Failed to load plugin: libpi_esimd_emulator.so
```

**Comment by vladimirlaz** [comment]
ICD loader for OpenCL or ze_loader are not displayed here

**Comment by smaslov-intel** [comment]
> ICD loader for OpenCL or ze_loader are not displayed here

Right, it only prints about loaded PI plugins shared libraries or dlerror reasons why anything wasn't loaded.

> layer of Level Zero loader might not be initialized properly and the device won't be detectable without any feedback from SYCL runtime.

The L0 plugin is linked to the L0 loader, so it is a must the L0 loader library is able to load for SYCL_PI_TRACE=1 to report success. If however, the loaded L0 loader doesn't function properly, then we might silently not see L0 devices. However, I do not see how dlerror can help.

Can someone describe the exact scenario, and maybe a reproducer?

**Comment by vladimirlaz** [comment]
In general, I would like to understand why I am missing some BE. There might be environment misconfiguration making SYCL RT miss some BE library. On Linux, I can use [strace](https://man7.org/linux/man-pages/man1/strace.1.html) to detect which library is missed. But it would be great if the missing library is clearly reported by SYCl RT (by default or when SYCL_PI_TRACE=1 is provided).
I think the request is not tight to dlopen but requires user-friendly reporting of misconfigurations.

**Comment by alexbatashev** [comment]
> On Linux, I can use [strace](https://man7.org/linux/man-pages/man1/strace.1.html) to detect which library is missed.

Right, but that essentially puts your application into debug mode and executes it step by step until `syscall` instruction is executed. I think that would be an overkill for runtime. And low-level runtimes and loaders are generally not very verbose about their failures.

**Comment by vladimirlaz** [comment]
> Member

I am not forcing the use of strace in RT. I would like to see user-friendly warding for missing highlevel BE libraries (loader/RT library)

**Comment by romanovvlad** [comment]
@vladimirlaz ping on the ask above.

**Comment by vladimirlaz** [comment]
it looks like the required info was added to xmain and is being upstreamed to intel/llvm: https://github.com/intel/llvm/pull/5628
